File: Codes/146_LRUcache/146.lru-缓存.go
/*
 * @Description:
 * @Autor: HTmonster
 * @Date: 2022-02-14 20:09:27
 */
/*
 * @lc app=leetcode.cn id=146 lang=golang
 *
 * [146] LRU 缓存
 */

// @lc code=start
type Node struct {
	Key   int
	Value int
	Prev  *Node
	Next  *Node
}

type LRUCache struct {
	DlinkHead *Node
	HashTable map[int]*Node
	Capacity  int
}

func Constructor(capacity int) LRUCache {
	// 创建双向链表
	head := Node{
		Key:   0,
		Value: 0,
	}
	tail := Node{
		Key:   0,
		Value: 0,
	}
	head.Next = &tail
	head.Prev = &tail
	tail.Next = &head
	tail.Prev = &head

	// 返回结构体
	return LRUCache{
		DlinkHead: &head,
		HashTable: make(map[int]*Node),
		Capacity:  capacity,
	}
}

func insertHead(head, node *Node) {
	//将节点插入链表的头部
	node.Next = head.Next
	head.Next.Prev = node
	head.Next = node
	node.Prev = head
}

func peakNode(node *Node) *Node {
	//摘出节点
	node.Prev.Next = node.Next
	node.Next.Prev = node.Prev
	return node
}

// func printLink(head *Node) {
// 	p := head
// 	for {
// 		if p.Next == head {
// 			break
// 		}
// 		fmt.Printf("%d-->", p.Key)
// 		p = p.Next
// 	}
// 	fmt.Printf("\n")
// }

func (this *LRUCache) Get(key int) int {
	fmt.Printf("get %d\n", key)
	// 检查是否存在
	node, ok := this.HashTable[key]
	// 不存在返回-1
	if !ok {
		// printLink(this.DlinkHead)
		return -1
	} else {
		// 存在
		// 摘出该节点
		node = peakNode(node)
		// 放到头部
		insertHead(this.DlinkHead, node)
		// printLink(this.DlinkHead)
		return node.Value
	}

}

func (this *LRUCache) Put(key int, value int) {
	fmt.Printf("put %d %d\n", key, value)
	//0.先检查是否存在
	node, ok := this.HashTable[key]
	if ok {
		//0.a 存在
		// 改值
		node.Value = value
		// 摘出节点
		node = peakNode(node)
	} else {
		//0.b 不存在
		// 创建节点
		node = &Node{
			Key:   key,
			Value: value,
		}
	}
	// 将节点插入头部
	insertHead(this.DlinkHead, node)
	// 存入hash表
	this.HashTable[key] = node
	// 检查并删除多余的元素
	fmt.Printf("%d!!!\n", len(this.HashTable))
	if len(this.HashTable) > this.Capacity {
		last := this.DlinkHead.Prev.Prev

		// 摘出节点
		last = peakNode(last)

		// 删除hash表
		delete(this.HashTable, last.Key)
		// 删除链表（自动）
	}
	// printLink(this.DlinkHead)
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * obj := Constructor(capacity);
 * param_1 := obj.Get(key);
 * obj.Put(key,value);
 */
// @lc code=end

File: Codes/146_LRUcache/146.lru-缓存.py
'''
Description: 
Autor: HTmonster
Date: 2022-02-14 19:07:34
'''
#
# @lc app=leetcode.cn id=146 lang=python3
#
# [146] LRU 缓存
#

# @lc code=start
class Node:

    def __init__(self,key=0,value=0):
        self.key = key
        self.value = value
        self.prev = None
        self.next=None


class LRUCache:
    # 两大数据结构
    #a. 双向链表  (用于保存顺序)
    DoubleLinkedList = None
    #b. hashtable (用于快速存取值)
    HashMap = None

    def printLink(self):
        print(self.HashMap)
        p=self.DoubleLinkedList
        while p.key!="tail":
            print("{}-->".format(p.key),end='')
            p=p.next
        print("tail")

    def __init__(self, capacity: int):
        # 初始化数据结构
        self.initDoubleLinkedList()
        self.capacity = capacity
        self.HashMap = {}
    
    def initDoubleLinkedList(self):
        head = Node(key="head")
        tail = Node(key="tail")
        head.next=tail
        head.prev=tail
        tail.next=head
        tail.prev=head
        self.DoubleLinkedList=head

        # self.printLink()

    def get(self, key: int) -> int:
        # print("get",key)
        # hash表中检查是否存在
        if key in self.HashMap.keys():
            # 挑战位置
            getNode = self.HashMap[key]
            # 摘出
            getNode.prev.next=getNode.next
            getNode.next.prev=getNode.prev
            
            # 插入头部
            self.insertHead(getNode)
            # self.printLink()
            return getNode.value
        else:
            # self.printLink()
            return -1

    def put(self, key: int, value: int) -> None:
        # print("put",key,value)
        #1. 先查找hash map
        if key in self.HashMap.keys():
            #2. 存在则修改值
            self.HashMap[key].value=value
            self.get(key)
        else:
            #3. 不存在则创建
            self.addNode(key, value)
        self.checkAndDelete()
        # self.printLink()
    
    def checkAndDelete(self):
        # 检查并删除多余的元素
        if len(self.HashMap.keys())>self.capacity:
            # 取出最后一个节点
            last_node=self.DoubleLinkedList.prev.prev
            # 删除hash表中的数据
            del self.HashMap[last_node.key]
            # 删除链表中的数据
            last_node.prev.next=last_node.next
            last_node.next.prev=last_node.prev
            del last_node
    
    def addNode(self,key,value):
        # 添加一个新节点
        node= Node(key=key,value=value)
        self.HashMap[key]=node
        self.insertHead(node)


    def insertHead(self,node):
        # 插入链表的头部
        node.next=self.DoubleLinkedList.next
        self.DoubleLinkedList.next.prev=node
        self.DoubleLinkedList.next=node
        node.prev=self.DoubleLinkedList

# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)
# @lc code=end
